{
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "metadata": {},
     "level": 1,
     "source": [
      "Analyze Inheritance Patterns"
     ]
    },
    {
     "cell_type": "code",
     "execution_count": null,
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "{'zygosity': 'hom', 'genes': 'HADH,HADHA,HADHB,ACAA1,ACAA2,EHHADH,ECHS1,HSD17B%', 'db': 'p7_platform.itmi_102_puzzle', 'member': 'all', 'platform': 'illumina', 'kav': '10', 'sample_id': 'all'}\n"
       ]
      }
     ],
     "input": [
      "#enter your user args as a comma-separated list with no spaces, or as 'all'\n",
      "args = {'member':'all', # this option not functional until Joe adds member column to variant tables\n",
      "        'genes':  'HADH,HADHA,HADHB,ACAA1,ACAA2,EHHADH,ECHS1,HSD17B%', \n",
      "        'sample_id':'all', \n",
      "        'kav':'10', \n",
      "        'zygosity':'hom', # this option not functional in illumina until Joe adds column\n",
      "        'platform':'illumina',\n",
      "        'db':'p7_platform.itmi_102_puzzle'\n",
      "        }\n",
      "\n",
      "print args"
     ],
     "language": "python",
     "prompt_number": 32
    },
    {
     "cell_type": "code",
     "execution_count": null,
     "metadata": {},
     "outputs": [],
     "input": [
      "##################################################\n",
      "# read in genes of interest and mark if wildcard #\n",
      "##################################################\n",
      "# split gene list by comma\n",
      "if args['genes'] is not None:\n",
      "    gene_args = args['genes'].replace(\"'\", \"\").split(',')\n",
      "else: \n",
      "    gene_args = []\n",
      "\n",
      "# mark any wildcards in gene list\n",
      "wildcards = []\n",
      "gene_list = []\n",
      "for gene in gene_args:\n",
      "    if gene.endswith('%'):\n",
      "        wildcards.append(gene)\n",
      "    elif len(gene_args) > 0:\n",
      "        gene_list.append(gene)\n",
      "\n",
      "###############\n",
      "# Build Query #\n",
      "###############\n",
      "# create statements for genes and wildcards\n",
      "if len(gene_list) > 0 and len(wildcards) < 1:\n",
      "    gene_statement = \"AND ens.gene_name IN ('\" + \"','\".join(map(str, gene_list)) + \"')\"\n",
      "elif len(wildcards) > 0 and len(gene_list) < 1:\n",
      "    gene_statement = 'AND ens.gene_name LIKE (' + \"','\".join(map(str, wildcards)) + \"')\"\n",
      "elif len(gene_list) > 0 and len(wildcards) > 0:\n",
      "   gene_statement = \"AND (ens.gene_name IN ('\" + \"','\".join(\n",
      "        map(str, gene_list)) + \"') OR ens.gene_name LIKE ('\" + \"',\" \\\n",
      "                                                           \"'\".join(map(str, wildcards)) + \"'))\"\n",
      "else:\n",
      "    gene_statement = \"\""
     ],
     "language": "python",
     "prompt_number": 33
    },
    {
     "cell_type": "code",
     "execution_count": null,
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n            SELECT p.sample_id, p.qual, p.filter, k.id as rsID, (k.alle_freq * 100) as kav_pct, k.alle_cnt as\n             kav_count, gene_name, p.chr, p.pos, p.ref, p.alt, p.gt,\n                   CASE  WHEN SUBSTRING(p.sample_id, -2) = '01' THEN 'M'\n                   WHEN SUBSTRING(p.sample_id, -2) = '02' THEN 'F'\n                   WHEN SUBSTRING(p.sample_id, -2) = '03' THEN 'NB' END as member,\n                   CONCAT(gene_name, ':', p.chr, ':', CAST(p.pos AS STRING)) as variant_id\n             FROM\n             (SELECT DISTINCT p.sample_id, p.qual, p.filter, ens.gene_name, p.chr, p.pos, p.ref, p.alt, p.gt\n             FROM public_hg19.ensembl_genes ens, p7_platform.itmi_102_puzzle AS p\n             WHERE ens.chromosome NOT LIKE 'H%'\n             AND (ens.gene_name IN ('HADH','HADHA','HADHB','ACAA1','ACAA2','EHHADH','ECHS1') OR ens.gene_name LIKE ('HSD17B%'))\n             AND ens.chromosome = p.chr\n             AND (p.pos >= ens.start AND p.pos <= ens.stop)\n             AND p.gt IS NOT NULL  ) AS p\n             LEFT JOIN /* +SHUFFLE */ public_hg19.kaviar k\n                   ON p.chr = k.chromosome\n                   AND p.pos = k.pos\n                   AND p.ref = k.ref\n                   AND p.alt = k.alt\n             WHERE (k.alle_freq < .10 OR k.alle_freq IS NULL)\n             \n"
       ]
      }
     ],
     "input": [
      "#build query\n",
      "if args['platform'] == 'illumina':\n",
      "    query = \"\"\"\n",
      "            SELECT p.sample_id, p.qual, p.filter, k.id as rsID, (k.alle_freq * 100) as kav_pct, k.alle_cnt as\n",
      "             kav_count, gene_name, p.chr, p.pos, p.ref, p.alt, p.gt,\n",
      "                   CASE  WHEN SUBSTRING(p.sample_id, -2) = '01' THEN 'M'\n",
      "                   WHEN SUBSTRING(p.sample_id, -2) = '02' THEN 'F'\n",
      "                   WHEN SUBSTRING(p.sample_id, -2) = '03' THEN 'NB' END as member,\n",
      "                   CONCAT(gene_name, ':', p.chr, ':', CAST(p.pos AS STRING)) as variant_id\n",
      "             FROM\n",
      "             (SELECT DISTINCT p.sample_id, p.qual, p.filter, ens.gene_name, p.chr, p.pos, p.ref, p.alt, p.gt\n",
      "             FROM public_hg19.ensembl_genes ens, {0:s} AS p\n",
      "             WHERE ens.chromosome NOT LIKE 'H%'\n",
      "             {1:s}\n",
      "             AND ens.chromosome = p.chr\n",
      "             AND (p.pos >= ens.start AND p.pos <= ens.stop)\n",
      "             AND p.gt IS NOT NULL  ) AS p\n",
      "             LEFT JOIN /* +SHUFFLE */ public_hg19.kaviar k\n",
      "                   ON p.chr = k.chromosome\n",
      "                   AND p.pos = k.pos\n",
      "                   AND p.ref = k.ref\n",
      "                   AND p.alt = k.alt\n",
      "             WHERE (k.alle_freq < .{2:s} OR k.alle_freq IS NULL)\n",
      "             \"\"\".format(args['db'], gene_statement, args['kav'])\n",
      "elif args['platform'] == 'cgi':\n",
      "    query = \"\"\"\n",
      "            SELECT p.sample_id, p.allele1varquality, p.totalreadcount, k.id as rsID, (k.alle_freq * 100) as kav_pct,\n",
      "            k.alle_cnt as kav_count, gene_name, p.chr, p.start, p.stop, p.ref, p.allele1seq, p.allele2seq, p.zygosity,\n",
      "            (CASE\n",
      "            WHEN SUBSTRING(p.sample_id, -1) = 'M' THEN 'M'\n",
      "            WHEN SUBSTRING(p.sample_id, -1) = 'F' THEN 'F'\n",
      "            WHEN SUBSTRING(p.sample_id, -2) = 'NB' THEN 'NB'\n",
      "            END) as member, CONCAT(gene_name, ':', p.chr, ':', CAST(p.start AS STRING), ':',\n",
      "            CAST(p.stop AS STRING)) as variant_id\n",
      "            FROM\n",
      "            (SELECT DISTINCT p.sample_id, p.allele1varquality, p.totalreadcount, ens.gene_name, p.chr, \\\n",
      "p.start, p.stop,\n",
      "             p.ref, p.allele1seq, p.allele2seq, p.zygosity\n",
      "             FROM public_hg19.ensembl_genes ens, {0:s} AS p\n",
      "             WHERE ens.chromosome NOT LIKE 'H%'\n",
      "             {1:s}\n",
      "             AND ens.chromosome = p.chr\n",
      "             AND (p.start >= ens.start AND p.stop <= ens.stop)\n",
      "             AND p.zygosity IS NOT NULL) AS p\n",
      "             LEFT JOIN /* +SHUFFLE */ public_hg19.kaviar k\n",
      "                ON p.chr = k.chromosome\n",
      "                AND (k.pos BETWEEN p.start and p.stop)\n",
      "                AND p.ref = k.ref\n",
      "                AND p.allele1seq = k.alt\n",
      "             WHERE (k.alle_freq < .{2:s} OR k.alle_freq IS NULL)\n",
      "            \"\"\".format(args['db'], gene_statement, args['kav'])\n",
      "else:\n",
      "    print \"Did you select illumina or cgi as your platform? Please check and try again.\"\n",
      "\n",
      "print query"
     ],
     "language": "python",
     "prompt_number": 34
    },
    {
     "cell_type": "code",
     "execution_count": null,
     "metadata": {},
     "outputs": [],
     "input": [
      "#################\n",
      "# load modules  #\n",
      "#################\n",
      "import argparse\n",
      "import sys\n",
      "import pandas as pd\n",
      "from impala.dbapi import connect\n",
      "from impala.util import as_pandas\n",
      "\n",
      "#stop irrelevant pandas warnings\n",
      "pd.options.mode.chained_assignment = None\n",
      "\n",
      "######################\n",
      "# connect to database #\n",
      "######################\n",
      "print \"Connecting to impala...\" + \"\\n\"\n",
      "\n",
      "#create database connection\n",
      "conn = connect(host='glados19', port=21050)\n",
      "cur = conn.cursor()\n",
      "\n",
      "#########################\n",
      "# Run Query on Impala ##\n",
      "#########################\n",
      "print \"Running query on impala...\" + \"\\n\"\n",
      "\n",
      "#execute query on impala\n",
      "cur.execute(query)\n",
      "query_df = as_pandas(cur)\n",
      "\n",
      "#check that results are returned and print preview\n",
      "if len(query_df) > 0:\n",
      "    print \"Results found, starting analysis...\"\n",
      "else:\n",
      "    print \"No results found\""
     ],
     "language": "python",
     "prompt_number": 35
    },
    {
     "cell_type": "code",
     "execution_count": null,
     "metadata": {},
     "outputs": [],
     "input": [
      "############################\n",
      "# create sets and subsets ##\n",
      "############################\n",
      "# create lists to store candidate variants\n",
      "hom_alt = []\n",
      "hom_ref = []\n",
      "comp_het = []\n",
      "mie = []\n",
      "missing = []\n",
      "\n",
      "# group by variant id to locate variants at same chr and pos\n",
      "by_variant = query_df.groupby('variant_id')\n",
      "\n",
      "# group variants by gene name\n",
      "by_gene = query_df.groupby('gene_name')"
     ],
     "language": "python"
    },
    {
     "cell_type": "code",
     "execution_count": null,
     "metadata": {},
     "outputs": [],
     "input": [
      "#######################################\n",
      "# find hom_alt and hom_ref candidates ##\n",
      "#######################################\n",
      "print \"Examining homozygous variants from het parents...\"\n",
      "\n",
      "for name, group in by_variant:\n",
      "    # find positions where both parents are het\n",
      "    if len(group[(group['member'] == 'M') & (group['gt'] == '0/1')]) > 0 \\\n",
      "        and (len(group[(group['member'] == 'F') & (group['gt'] == '0/1')]) > 0):\n",
      "        # if the newborn is hom_ref\n",
      "        if (len(group[(group['member'] == 'NB') & (group['gt'] == '0/0')]) > 0):\n",
      "            # mark as hom_ref\n",
      "            group['var_type'] = \"hom_ref\"\n",
      "            hom_ref.append(group)\n",
      "        # if the newborn is hom_alt\n",
      "        elif (len(group[(group['member'] == 'NB') & (group['gt'] == '1/1')]) > 0):\n",
      "            group['var_type'] = \"hom_alt\"\n",
      "            hom_alt.append(group)\n",
      "if len(hom_alt) > 0:\n",
      "    print \"{0:s} Homozygous alt variants found. \\n\".format(len(hom_alt))\n",
      "else:\n",
      "    print \"No homozygous alt variants found.  \\n\"\n",
      "    \n",
      "if len(hom_ref) > 0:\n",
      "    print \"{0:s} Homozygous ref variants found. \\n\".format(len(hom_ref))\n",
      "else:\n",
      "    print \"No homozygous ref variants found.  \\n\""
     ],
     "language": "python"
    },
    {
     "cell_type": "code",
     "execution_count": null,
     "metadata": {},
     "outputs": [],
     "input": [
      "###################\n",
      "# find comp_hets ##\n",
      "###################\n",
      "print \"Looking for compound heterozygosity...\"\n",
      "\n",
      "for name, group in by_gene:\n",
      "    # if there are variants at more than one position and there are nb het variants at diff pos\n",
      "    if group.pos.nunique() > 1 and (len(group[((group['member'] == 'NB') & (group['gt'] == '0/1'))].pos) > 1):\n",
      "        # pull out nb het positions\n",
      "        nb_hets = group[((group['member'] == 'NB') & (group['gt'] == '0/1'))]\n",
      "        # add parent positions for the gene\n",
      "        moms = nb_hets.append(group[((group['member'] == \"M\") & (group['variant_id'].isin(nb_hets['variant_id'])))],\n",
      "                              ignore_index=True)\n",
      "        all_vars = moms.append(group[((group['member'] == \"F\") & (group['variant_id'].isin(nb_hets['variant_id'])))],\n",
      "                               ignore_index=True)\n",
      "        # if there are matching variants from both parents, mark as comp-het\n",
      "        if (len(all_vars[(all_vars['member'] == 'F')]) > 0) and (len(all_vars[(all_vars['member'] == 'M')]) > 0):\n",
      "            all_vars['var_type'] = \"comp_het\"\n",
      "            comp_het.append(all_vars)\n",
      "            \n",
      "if len(comp_het) > 0:\n",
      "    print \"{0:s} Compound het variants found. \\n\".format(len(comp_het))\n",
      "else:\n",
      "    print \"No compound het variants found.  \\n\""
     ],
     "language": "python"
    },
    {
     "cell_type": "code",
     "execution_count": null,
     "metadata": {},
     "outputs": [],
     "input": [
      "#############\n",
      "# find MIE ##\n",
      "#############\n",
      "print \"Searching for MIE....\"\n",
      "\n",
      "# het variants\n",
      "for name, group in by_variant:\n",
      "    #if newborn is het\n",
      "    if (len(group[(group['member']=='NB') & (group['gt']=='0/1')]) > 0 ):\n",
      "        # and both parents at this position are homozygous\n",
      "        if (len(group[((group['member']=='M') & ((group['gt'] == '0/0')| (group['gt'] == '1/1')))]) > 0) & \\\n",
      "         (len(group[((group['member']=='F') & ((group['gt'] == '0/0')| (group['gt'] == '1/1')))]) > 0):\n",
      "            if (len(group[(group['member'] == 'F')]) > 0) & (len(group[(group['member'] == 'M')]) > 0):\n",
      "                group['var_type'] = \"mie\"\n",
      "                mie.append(group)\n",
      "            else:\n",
      "                group['var_type'] = \"missing_parent\"\n",
      "                missing.append(group)\n",
      "        # or newborn variant is not from either parent\n",
      "        elif (group[(group['member'] == 'NB')].alt.any() != (group[(group['member'] == 'F')].alt.any() \\\n",
      "                                                                     or group[(group['member'] == 'M')].alt.any())):\n",
      "            if (len(group[(group['member'] == 'F')]) > 0) & (len(group[(group['member'] == 'M')]) > 0):\n",
      "                group['var_type'] = \"mie\"\n",
      "                mie.append(group)\n",
      "            else:\n",
      "                group['var_type'] = \"missing_parent\"\n",
      "                missing.append(group)\n",
      "\n",
      "# hom variants\n",
      "for name, group in by_variant:\n",
      "    # if newborn is homozygous\n",
      "    if (len(group[(group['member']=='NB') & (group['gt']=='1/1')]) > 0 ):\n",
      "        # and only one parent is het for alt allele\n",
      "        if ((len(group[( ((group['member']=='F') & (group['gt']=='1/1')) & ((group['member']=='M') & (group['gt']!='1/1')) )]) > 0 ) | (len(group[( ((group['member']=='M') & (group['gt']=='1/1')) & ((group['member']=='F') & (group['gt']!='1/1')) )]) > 0 )):\n",
      "            if (len(group[(group['member'] == 'F')]) > 0) & (len(group[(group['member'] == 'M')]) > 0):\n",
      "                group['var_type'] = \"mie\"\n",
      "                mie.append(group)\n",
      "            else:\n",
      "                group['var_type'] = \"missing_parent\"\n",
      "                missing.append(group)\n",
      "        #or nb variant is not from either parent\n",
      "        elif (group[(group['member'] == 'NB')].alt.any() != (group[(group['member'] == 'F')].alt.any() \\\n",
      "                                                                     or group[(group['member'] == 'M')].alt.any())):\n",
      "            if (len(group[(group['member'] == 'F')]) > 0) & (len(group[(group['member'] == 'M')]) > 0):\n",
      "                group['var_type'] = \"mie\"\n",
      "                mie.append(group)\n",
      "            else:\n",
      "                group['var_type'] = \"missing_parent\"\n",
      "                missing.append(group)\n",
      "                \n",
      "if len(mie) > 0:\n",
      "    print \"{0:s} MIE variants found. \\n\".format(len(mie))\n",
      "else:\n",
      "    print \"No MIE variants found.  \\n\"\n",
      "    \n",
      "if len(missing) > 0:\n",
      "    print \"{0:s} variants found with missing parent info. \\n\".format(len(hom_alt))\n",
      "else:\n",
      "    print \"No variants found with missing parent info.  \\n\""
     ],
     "language": "python"
    },
    {
     "cell_type": "code",
     "execution_count": null,
     "metadata": {},
     "outputs": [],
     "input": [
      "###########################\n",
      "# merge and save results ##\n",
      "###########################\n",
      "print \"Saving results to current working directory....\"\n",
      "\n",
      "#create list of groupby result lists and merge\n",
      "groups = hom_alt + hom_ref + comp_het + mie + missing\n",
      "results = pd.concat(groups, axis=0, join='outer')\n",
      "\n",
      "#save to csv\n",
      "results.to_csv('rare_variant_results.csv', header=True, encoding='utf-8')\n",
      "\n",
      "#close all the things\n",
      "cur.close()\n",
      "sys.exit()"
     ],
     "language": "python"
    }
   ]
  }
 ],
 "cells": [],
 "metadata": {},
 "nbformat": 3,
 "nbformat_minor": 0
}