---
title: "ucsc_illumina_queries"
output: md_document
---

#Sample queries for ucsc gene annotations

##Find all variants in ucsc gene regions
The following query can be used to determine what gene regions a variant lies in and to create a subset for downstream annotation for finding amino acid changes. 

Update or remove the following variables, as needed:  
- ill.filter: filter out variants that did not pass filtering by the variant caller  
- ill.qual: filter out variants below a certain quality score  
- ucsc.strand: Find only matches on the forwad strand, remove for both, or "-" for reverse


```sql
SELECT ill.sample_id as ill_sample_id, ill.chromosome as ill_chrom, 
ill.position as ill_pos,ill.ref as ill_ref, ill.alt as ill_alt, ill.id as ill_rsID, 
ucsc.name as ucsc_id, SUBSTR(ucsc.chrom,4) as ucsc_chrom, ucsc.strand as ucsc_strand, 
ucsc.txstart as ucsc_txstart, ucsc.txend as ucsc_txend, ucsc.cdsstart as ucsc_cdsstart, 
ucsc.cdsend as ucsc_cdsend, ucsc.exonstarts as ucsc_exonstart, ucsc.exonends as ucsc_exonend, 
ucsc.proteinid as ucsc_proteinid, ucsc.alignid as ucsc_alignid
FROM p7dev.illumina_test as ill, public_hg19.ucsc as ucsc
WHERE ill.filter = "PASS"  
AND ucsc.strand = "+"
AND ill.chromosome = SUBSTR(ucsc.chrom,4) 
AND ill.position BETWEEN ucsc.txstart AND ucsc.txend  
```

##Find specific genes in the ucsc gene subset created above
Update or remove the following variables, as needed:  
- ill.filter: filter out variants that did not pass filtering by the variant caller  
- ill.qual: filter out variants below a certain quality score  
- ucsc.strand: Find only matches on the forwad strand, remove for both, or "-" for reverse
- ucsc.name: enter each gene of interest as a list, with each gene in quotes, comma-separated  

```sql
SELECT ill.sample_id as ill_sample_id, ill.chromosome as ill_chrom, 
ill.position as ill_pos,ill.ref as ill_ref, ill.alt as ill_alt, ill.id as ill_rsID, 
ucsc.name as ucsc_id, SUBSTR(ucsc.chrom,4) as ucsc_chrom, ucsc.strand as ucsc_strand, 
ucsc.txstart as ucsc_txstart, ucsc.txend as ucsc_txend, ucsc.cdsstart as ucsc_cdsstart, 
ucsc.cdsend as ucsc_cdsend, ucsc.exonstarts as ucsc_exonstart, ucsc.exonends as ucsc_exonend, 
ucsc.proteinid as ucsc_proteinid, ucsc.alignid as ucsc_alignid
FROM p7dev.illumina_test as ill, public_hg19.ucsc as ucsc
WHERE ill.filter = "PASS"  
AND ucsc.strand = "+"
AND ill.chromosome = SUBSTR(ucsc.chrom,4) 
AND ill.position BETWEEN ucsc.txstart AND ucsc.txend  
AND ucsc.name IN ("uc009xhs.1", "uc010qcb.1", "uc001ioc.1", "uc001ioc.1", "uc010qcb.1")
```

##Find a specific gene in the ucsc gene subset 
Update or remove the following variables, as needed:  
- ill.filter: filter out variants that did not pass filtering by the variant caller  
- ill.qual: filter out variants below a certain quality score  
- ucsc.strand: Find only matches on the forwad strand, remove for both, or "-" for reverse
- ucsc.name: Enter gene of interest, surrounded by quotes

```sql
SELECT ill.sample_id as ill_sample_id, ill.chromosome as ill_chrom, 
ill.position as ill_pos,ill.ref as ill_ref, ill.alt as ill_alt, ill.id as ill_rsID, 
ucsc.name as ucsc_id, SUBSTR(ucsc.chrom,4) as ucsc_chrom, ucsc.strand as ucsc_strand, 
ucsc.txstart as ucsc_txstart, ucsc.txend as ucsc_txend, ucsc.cdsstart as ucsc_cdsstart, 
ucsc.cdsend as ucsc_cdsend, ucsc.exonstarts as ucsc_exonstart, ucsc.exonends as ucsc_exonend, 
ucsc.proteinid as ucsc_proteinid, ucsc.alignid as ucsc_alignid,
CASE 
WHEN ill.position BETWEEN ucsc.cdsstart AND ucsc.cdsend 
THEN 'CDS' 
WHEN ill.position BETWEEN ucsc.txstart AND ucsc.txend
THEN 'transcript'
ELSE NULL
             END AS ucsc_feature
FROM p7dev.illumina_test as ill, public_hg19.ucsc as ucsc
WHERE ill.filter = "PASS"  
AND ucsc.strand = "+"
AND ill.chromosome = SUBSTR(ucsc.chrom,4) 
AND ill.position BETWEEN ucsc.txstart AND ucsc.txend  
AND ucsc.name = ("uc009xhs.1")
```
##Find variants in located in coding regions
Use this query to find all variants located in a region annotated by ucsc as coding. 

Update or remove the following variables, as needed:  
- ill.filter: filter out variants that did not pass filtering by the variant caller  
- ill.qual: filter out variants below a certain quality score  
- ucsc.strand: Find only matches on the forwad strand, remove for both, or "-" for reverse
- ucsc_table.ucsc_feature: enter either "CDS" or "transcript" surrounded by quotes

```sql
WITH ucsc_table AS (
SELECT ill.sample_id as ill_sample_id, ill.chromosome as ill_chrom, 
ill.position as ill_pos,ill.ref as ill_ref, ill.alt as ill_alt, ill.id as ill_rsID, 
ucsc.name as ucsc_id, SUBSTR(ucsc.chrom,4) as ucsc_chrom, ucsc.strand as ucsc_strand, 
ucsc.txstart as ucsc_txstart, ucsc.txend as ucsc_txend, ucsc.cdsstart as ucsc_cdsstart, 
ucsc.cdsend as ucsc_cdsend, ucsc.exonstarts as ucsc_exonstart, ucsc.exonends as ucsc_exonend, 
ucsc.proteinid as ucsc_proteinid, ucsc.alignid as ucsc_alignid,
CASE 
WHEN ill.position BETWEEN ucsc.cdsstart AND ucsc.cdsend 
THEN 'CDS' 
WHEN ill.position BETWEEN ucsc.txstart AND ucsc.txend
THEN 'transcript'
ELSE NULL
             END AS ucsc_feature
FROM p7dev.illumina_test as ill, public_hg19.ucsc as ucsc
WHERE ill.filter = "PASS"  
AND ucsc.strand = "+"
AND ill.chromosome = SUBSTR(ucsc.chrom,4) 
AND ill.position BETWEEN ucsc.txstart AND ucsc.txend 
  )
  SELECT * FROM ucsc_table
  WHERE ucsc_table.ucsc_feature = "CDS"
```

##Find variant located in eiher one feature or another
Use this query to find all variants located in either one region, such as coding, or another, such as exon. 

Update or remove the following variables, as needed:  
- ill.filter: filter out variants that did not pass filtering by the variant caller  
- ill.qual: filter out variants below a certain quality score  
- ucsc.strand: Find only matches on the forwad strand, remove for both, or "-" for reverse
- ucsc_table.ucsc_feature: enter either "CDS" or "transcript" surrounded by quotes

```sql
WITH ucsc_table AS (
SELECT ill.sample_id as ill_sample_id, ill.chromosome as ill_chrom, 
ill.position as ill_pos,ill.ref as ill_ref, ill.alt as ill_alt, ill.id as ill_rsID, 
ucsc.name as ucsc_id, SUBSTR(ucsc.chrom,4) as ucsc_chrom, ucsc.strand as ucsc_strand, 
ucsc.txstart as ucsc_txstart, ucsc.txend as ucsc_txend, ucsc.cdsstart as ucsc_cdsstart, 
ucsc.cdsend as ucsc_cdsend, ucsc.exonstarts as ucsc_exonstart, ucsc.exonends as ucsc_exonend, 
ucsc.proteinid as ucsc_proteinid, ucsc.alignid as ucsc_alignid,
CASE 
WHEN ill.position BETWEEN ucsc.cdsstart AND ucsc.cdsend 
THEN 'CDS' 
WHEN ill.position BETWEEN ucsc.txstart AND ucsc.txend
THEN 'transcript'
ELSE NULL
             END AS ucsc_feature
FROM p7dev.illumina_test as ill, public_hg19.ucsc as ucsc
WHERE ill.filter = "PASS"  
AND ucsc.strand = "+"
AND ill.chromosome = SUBSTR(ucsc.chrom,4) 
AND ill.position BETWEEN ucsc.txstart AND ucsc.txend 
  )
  SELECT * FROM ucsc_table
  WHERE (ucsc_table.ucsc_feature= "CDS" or ucsc_table.ucsc_feature = "transcript")
```

##Find variants that are not located in ucsc gene regions
This query can be used downstream to locate intergenic variants by finding the two flanking genes, and distances between the variants and the flanking genes.

Update or remove the following variables, as needed:  
- ill.filter: filter out variants that did not pass filtering by the variant caller  
- ill.qual: filter out variants below a certain quality score  

```sql
SELECT ill.* 
FROM
  p7dev.illumina_test as ill
  LEFT JOIN public_hg19.ucsc as ucsc
    ON ill.chromosome = SUBSTR(ucsc.chrom,4) 
    AND ill.position BETWEEN ucsc.txstart AND ucsc.txend
WHERE
  ill.filter = "PASS" 
  AND ucsc.chrom IS NULL
```



library(VariantAnnotation)
getMethods(locateVariants)

