---
title: "testing_illumina_ensmebl.rmd"
output: html_document
---
###Purpose: 
Testing ensembl gene annotation queries for illumina variants. 

###Test plan
The first 100000 rows of  p7_ptb.illumina_variant and public_hg19.ensembl_genes will be saved as p7dev.illumina_test and p7dev.ensembl_test on impala and run through tests to compare results between R and impala queries. 

After initial check to make sure tables in R and impala are both 100,000 rows:

1. Run coding gene annotation query on impala and R 
2. Compare counts of total records returned  
3. Compare counts of unique zygosity entries  
4. Compare counts of unique variant types returned 
5. Compare counts of unique gene names reported  
6. Compare random sample of results between R, impala and dbSNP

###Finding variants on impala
The following query was run to count the number of anntoated genes returned: 

```sql 
SELECT COUNT(*)
FROM p7dev.illumina_test as vcf, p7dev.ensembl_test as ens
WHERE vcf.filter = "PASS"  
AND vcf.chromosome = ens.chromosome  
AND vcf.position BETWEEN ens.start AND ens.stop  
```
450 rows were returned. 

###Downloading variants into R
The following query was run to return results for later comparison with R results: 
```sql
SELECT vcf.sample_id as vcf_sample_id, vcf.chromosome as vcf_chrom, vcf.position as vcf_pos,vcf.ref as vcf_ref, vcf.alt as vcf_alt, vcf.id as vcf_rsID, ens.start as ensembl_start, ens.stop as ensembl_end, ens.feature, ens.gene_name as ensembl_gene_name, ens.gene_id as ensembl_geneid, ens.gene_biotype as ensembl_gene_biotype, ens.transcript_name as ensembl_tx_name,ens.transcript_id as ensembl_trans_id, ens.exon_id as ensembl_exonid, ens.strand as ensembl_strand
FROM p7dev.illumina_test as vcf, p7dev.ensembl_test as ens
WHERE vcf.filter = "PASS"  
AND vcf.chromosome = ens.chromosome  
AND vcf.position BETWEEN ens.start AND ens.stop  
```
This file was then read into R to ensure validity of queries used for tesing: 

```{r read_results}
#read in results table
ens_results = read.csv("D:/Documents/GitHub/impala_scripts/queries/testing/ensembl/ensembl_query_result.csv", header=TRUE)
head(ens_results)
```

###Reading data into R for comparison

The subset tables were read into R: 

```{r read_in_subset}
suppressMessages(library(RODBC))

#connect using the DSN name you created on your machine
conn <- odbcConnect("Impala DSN")

vcf = sqlFetch(conn, "p7dev.illumina_test")
head(vcf)
dim(vcf)
ens = sqlFetch(conn, "p7dev.ensembl_test")
head(ens)
dim(ens)
```

The following was run in R to find overalps bewteen cgi variants and ensembl genes: 
```{r count_features_R}
suppressMessages(library(GenomicRanges))
suppressMessages(library(IRanges))

##subset for only vcf rows that passed filtering
vcf_pass = vcf[which(vcf$filter == "PASS"),]
        
#create GRanges objects for comparison
vars = with(vcf_pass, GRanges(seqnames=vcf_pass$chromosome, 
                         ranges=IRanges(vcf_pass$pos, width=1, 
                                        names=paste(vcf_pass$sample_id, ":",vcf_pass$ref, ">", vcf_pass$alt, sep="")),rsID=vcf_pass$id, filter=vcf_pass$filter, qual=vcf_pass$qual, sample_id=vcf_pass$sample_id))

genes = with(ens, GRanges(seqnames=ens$chromosome, 
                         ranges=IRanges(ens$start, ens$stop,names=ens$gene_id),
                                        feature=ens$feature, trans_id=ens$transcript_id, gene_id=ens$gene_id, gene_name=ens$gene_name, trans_name=ens$transcript_name, exon=ens$exon_id))

##find variants that fall in gene regions
match = suppressWarnings(findOverlaps(vars, genes))

##turn results into a data frame
match.df = data.frame(sample_id = as.character(mcols(vars)$sample_id)[queryHits(match)],
                      chrom = seqnames(vars)[queryHits(match)],
                      pos = as.data.frame(ranges(vars))$start[queryHits(match)],
                      variant = as.character(lapply(strsplit(as.character(names(vars)[queryHits(match)]), ":"), function(x) x[2])),
                      rsId = as.character(mcols(vars)$rsID)[queryHits(match)],
                      gene_name = mcols(genes)$gene_name[subjectHits(match)],
                      gene_id = names(genes)[subjectHits(match)],
                      trans_name = mcols(genes)$trans_name[subjectHits(match)],
                      trans_id = mcols(genes)$trans_id[subjectHits(match)],
                      exon_id = mcols(genes)$exon[subjectHits(match)],
                      feature = mcols(genes)$feature[subjectHits(match)],
                      stringsAsFactors=F)

#450 rows, same as impala query                                  
head(match.df)
dim(match.df)   

```

###compare Impala Results file and R generated results
```{r compare_results}
#add variant column to impala results
ens_results$variant = paste(ens_results$vcf_ref, ">", ens_results$vcf_alt, sep="")
#coerce to same structure as match.df for comparison
ens.df = ens_results[c(1,2,3,17,6,10,11,13,14,15,9)]
colnames(ens.df) = colnames(match.df)

#order both data frames for matching
ens.df = ens.df[with(ens.df, order(sample_id, chrom, pos, variant, gene_name, gene_id, trans_name, trans_id, exon_id, feature)),]
match.df = match.df[with(match.df, order(sample_id, chrom, pos, variant, gene_name, trans_name, trans_id, exon_id, feature)),]

#coercing columns to same class for comparison
i = sapply(ens.df, is.factor)
ens.df[i] = lapply(ens.df[i], as.character)

j = sapply(match.df, is.factor)
match.df[j] = lapply(match.df[j], as.character)

#removing rownames for matching
rownames(match.df)= NULL
rownames(ens.df) = NULL

#rsID column is "NULL" in one table, "NA" in the other
ens.df$rsId = gsub("NULL", "NA", ens.df$rsId)

#compare data frames
which(match.df != ens.df, arr.ind=TRUE)

```

No results returned means that the results are identical. 

###Count unique gene features returned
Imapala: 

This following query was run on impala, showing two feature categories "CDS" and "exon": 
```sql
SELECT COUNT(DISTINCT ens.feature)
FROM p7dev.illumina_test as vcf, p7dev.ensembl_test as ens
WHERE vcf.filter = "PASS"  
AND vcf.chromosome = ens.chromosome  
AND vcf.position BETWEEN ens.start AND ens.stop  
```

R on impala results: 

The following R query run on the impala results generated above shows that there are two features found in the results, "CDS" and "exon". 

```{r count_results_features}
str(ens_results$feature)
```

R results: 
Results from the R-generated data also show to features, "CDS" and "exon". 

```{r count_R_features}
unique(match.df$feature)

```

###Compare random sample of genes returned
First, we'll grab a random sample of genes from R and examine them on each platform: 
```{r rand_genes}
#rand_genes = sample(unique(match.df$gene_name), 5)
#the above generated the following list, saving as fixed ("RMRPP1","PPIAP13","NDST2","RP11-574K11.8","RPL39P25")
rand_genes = c("RMRPP1","PPIAP13","NDST2","RP11-574K11.8","RPL39P25")
```

How many rows are returned for these genes in R? 
```{r count_rand_genes}

rand_match = match.df[(match.df$gene_name %in% rand_genes),]
dim(rand_match)
rand_match

```

How many of these genes are in impala results? 

```{r count_rand_impala_results}

match_imp_results = ens.df[(ens.df$gene_name %in% rand_genes),]
dim(match_imp_results)
match_imp_results

##are the gene also the same in the imapla results?
which(rand_match != match_imp_results, arr.ind=TRUE)
#returns 0 mismatches
```

12 identical results are returned on impala: 
```sql
SELECT COUNT(*)
FROM p7dev.illumina_test as vcf, p7dev.ensembl_test as ens
WHERE vcf.filter = "PASS"  
AND vcf.chromosome = ens.chromosome  
AND vcf.position BETWEEN ens.start AND ens.stop 
AND ens.gene_name IN ("RMRPP1","PPIAP13","NDST2","RP11-574K11.8","RPL39P25")
```

###Query to find variants not in ensembl annotated regions
Results of this query can be used to locate intergenic variants; downstream analysis can be used to find the two flanking genes, and distances between the variants and the flanking genes.

The following query was run on sql, and the results were downloaded for analysis in R: 
```sql
SELECT vcf.* 
FROM
  p7dev.illumina_test as vcf
  LEFT JOIN p7dev.ensembl_test as ens
    ON vcf.chromosome = ens.chromosome
    AND vcf.position BETWEEN ens.start AND ens.stop
WHERE
  vcf.filter = "PASS" 
  AND ens.chromosome IS NULL
```

###Examining random sample of intergenic variants
A random sample of intergenic variants was compared annoated in R using the bioconductor hg19 database: 


```{r load_intergenic}
intergenic_csv = read.csv("D:/Documents/GitHub/impala_scripts/queries/testing/ensembl/intergenic_test.csv", header=TRUE)
head(intergenic_csv)

#pull random sample and assign to variable
rand_ill_intergenic = intergenic_csv[sample(unique(c(intergenic_csv$sample_id)), 5),]
rand_ill_intergenic

suppressMessages(library(VariantAnnotation))
suppressMessages(library(AnnotationHub))
suppressMessages(library(TxDb.Hsapiens.UCSC.hg19.knownGene))

#load referenece genome from ucsc knowngene db
txdb_hg19 <- TxDb.Hsapiens.UCSC.hg19.knownGene


#create granges object from random sample of intergenics
not_genes = with(rand_ill_intergenic, GRanges(seqnames=chromosome, 
                              ranges=IRanges(position, width=1, 
                              names=paste(sample_id, ":",ref, ">", alt, sep="")),
                              rsID = id,
                              sample_id = sample_id))

#chr names don't match
head(seqlevels(txdb_hg19))
seqlevels(not_genes)
#matching chr names
seqlevels(not_genes) <- paste0("chr", seqlevels(not_genes))
not_genes

#assign genome
genome(not_genes) = "hg19"

#confirm that random sample is intergenic
locateVariants(not_genes, txdb_hg19, AllVariants())

```

