---
title: "find_variants_ben"
output: pdf_document
---
# Purpose: 
Pull git from home for the intro. 

#Query
This query was run to find variants in regions of interest: 
```sql
WITH ens AS (
 SELECT DISTINCT chromosome as chr, start, stop, gene_name
   FROM public_hg19.ensembl_genes
  WHERE (gene_name IN ( 'HADH', 'HADHA', 'HADHB', 'ACAA1',
                        'ACAA2', 'EHHADH', 'ECHS1')
     OR gene_name LIKE 'HSD17B%')
    AND chromosome NOT LIKE "H%"
)
SELECT p.*, (k.alle_freq * 100) as kav_freqPct, k.alle_cnt as kav_count, 
    (CASE  
      WHEN SUBSTRING(p.sample_id, -2) = '01'
      THEN 'M'
      WHEN SUBSTRING(p.sample_id, -2) = '02'
      THEN 'F'
      WHEN SUBSTRING(p.sample_id, -2) = '03'
      THEN 'NB'
      END) as member      
  FROM
(SELECT DISTINCT p.sample_id, p.chr, p.pos, p.id,
                 p.ref, p.alt, p.gt, p.qual, p.filter,
                 ens.gene_name
   FROM ens, p7_ptb.itmi_102_puzzle p
  WHERE p.chr = ens.chr
    AND p.pos >= ens.start AND p.pos <= ens.stop
    AND p.gt IS NOT NULL
) AS p
LEFT JOIN public_hg19.kaviar k
       ON p.chr = k.chromosome
      AND p.pos = k.pos
      AND p.ref = k.ref
      AND p.alt = k.alt
      AND k.alle_freq < .10
```

#Search for homozygous alt alleles

```{r hom_alt}
suppressMessages(library(readr))

#read in query results from impala
query = read_csv("~/GitHub/impala_scripts/queries/brady_requests/ben_query.csv")

#split apart data frame by member
mom = query[which(query$member == "M"),]
dad = query[which(query$member == "F"),]
nb = query[which(query$member == "NB"),]

#add gene:chr:pos id to each df
mom$variant_id = paste(mom$gene_name, mom$chr, mom$pos, sep=":")
dad$variant_id = paste(dad$gene_name, dad$chr, dad$pos, sep=":")
nb$variant_id = paste(nb$gene_name, nb$chr, nb$pos, sep=":")

#add id for alt allele
mom$alt_id = paste(mom$chr, mom$pos, mom$alt, sep=":")
dad$alt_id = paste(dad$chr, dad$pos, dad$alt, sep=":")
nb$alt_id = paste(nb$chr, nb$pos, nb$alt, sep=":")

#######################
#find homozygous alt ##
#######################

#subset for zygosity congruent with hom alt mutations
hom_alt_nb = nb[which(nb$gt == "1/1"),]
het_mom = mom[which(mom$gt == "0/1"),]
het_dad = dad[which(dad$gt == "0/1"),]

#merge together to match by gene, chr and position
nb_mom_hom_alt = merge(hom_alt_nb, het_mom, by=c("gene_name", "chr", "pos"))
#get rid of unnecessary columns
nb_mom_hom_alt = nb_mom_hom_alt[,c(1:3,5,9:12,6,7,17)]

#merge with dad
hom_alt = merge(nb_mom_hom_alt, het_dad, by=c("gene_name", "chr", "pos"))

#get rid of unnecessary columns
colnames(hom_alt)
hom_alt = hom_alt[,c(1:11,15)]
colnames(hom_alt) = c("gene_name", "chr", "pos", "rsID", "qual", "filter", "kav_freqPct", 
                      "kav_count", "ref", "nb_alt", "m_alt", "f_alt")

hom_alt.df = hom_alt[which((hom_alt$nb_alt == hom_alt$m_alt) & (hom_alt$nb_alt == hom_alt$f_alt)),]
hom_alt.df
```

##Search for homozygous ref alleles

```{r hom_ref}
#############################
#find homozygous reference ##
#############################
hom_ref_nb = nb[which(nb$gt == "0/0"),]

#merge hom rec nb with het mom
hom_ref_nb_mom = merge(hom_ref_nb, het_mom, by=c("gene_name", "chr", "pos"))
#get rid of unnecessary columns
hom_ref_nb_mom = hom_ref_nb_mom[,c(1:3,5,9:12,6,7,17)]

#merge with dad
hom_ref.df = merge(hom_ref_nb_mom, het_dad, by=c("gene_name", "chr", "pos"))
#get rid of unnecessary columns
hom_ref.df = hom_ref.df[,c(1:11,15)]
colnames(hom_ref.df) = c("gene_name", "chr", "pos", "rsID", "qual", "filter", "kav_freqPct", 
                      "kav_count", "ref", "nb_alt", "m_alt", "f_alt")
hom_ref.df

```

##Search for compound het alleles

```{r comp_het}
#####################
#find compound het ##
#####################
suppressMessages(library(stats))

#subset for possible zygosity for compound het
het_nb = nb[which(nb$gt == "0/1"),]
comp_mom = mom[which(mom$gt == "0/1" | mom$gt == "1/1"),]
comp_dad = dad[which(dad$gt == "0/1" | dad$gt == "1/1"),]

#order data frames for split by function
het_nb = het_nb[order(het_nb$gene_name),]
comp_mom = comp_mom[order(comp_mom$gene_name),]
comp_dad = comp_dad[order(comp_dad$gene_name),]

#function to check for comp_hets
find_comphet = function(x){
  #check if more than one variant found per gene
  if (length(unique(x$pos)) >1 ) {
    #find matching variants from the parents
    mom_vars = comp_mom[grep(paste(x$variant_id, collapse="|"), comp_mom$variant_id),]
    dad_vars = comp_dad[grep(paste(x$variant_id, collapse="|"), comp_dad$variant_id),]
    #check that parent variants are different and at different positions
    if (dim(mom_vars)[1] > 1 & dim(dad_vars)[1] > 1){
      #find variants that are same in mom and dad
      same = merge(mom_vars, dad_vars, by="variant_id")
      #check that both mom and dad have variants that are different
      if (dim(dad_vars[!(dad_vars$variant_id %in% same$variant_id),])[1] >= 1 &
            dim(mom_vars[!(mom_vars$variant_id %in% same$variant_id),])[1] >= 1){
        #pull out comp hets
        comp_hets.df = rbind(het_nb[grep(paste(mom_vars[!(mom_vars$variant_id %in% same$variant_id),]$variant_id,
                                               collapse="|"),  
        het_nb$variant_id),],
        het_nb[grep(paste(dad_vars[!(dad_vars$variant_id %in% same$variant_id),]$variant_id, collapse="|"),
        het_nb$variant_id),],
        mom_vars[!(mom_vars$variant_id %in% same$variant_id),],
        dad_vars[!(dad_vars$variant_id %in% same$variant_id),])
        comp_hets.df[order(comp_hets.df$gene_name, comp_hets.df$chr, comp_hets.df$pos),]
            }else
              print("Same variants found in both parents")     
      } else
        paste("Variants only found in one parent.", sep=" ")
    } else
      paste("Only one variant found in gene.")
}

##apply function to newborn comp het candidates
comp_het_list = by(het_nb, het_nb$gene_name, find_comphet)

comp_het.df = do.call(rbind, comp_het_list)
comp_hets = comp_het.df[which(comp_het.df$sample_id != "Variants only found in one parent." 
                  & comp_het.df$sample_id != "Same variants found in both parents"
            & comp_het.df$sample_id != "Variants only found in one parent."
            & comp_het.df$sample_id != "Only one variant found in gene."),]
rownames(comp_hets)  = NULL

comp_hets

```

##Search for Mendelian Inheritance Errors

```{r find_mie}

#subset for newborns not found above
comp_hets$variant_type = "comp_het"
hom_ref.df$variant_type = "hom_ref"
hom_alt.df$variant_type = "hom_alt"

##go back and get all the columns for hom variants ##

#data frame homozygous variants
homozygous = rbind(hom_ref.df, hom_alt.df)
homozygous$variant_id = paste(homozygous$gene_name, homozygous$chr, homozygous$pos, sep=":")

#find nb variants not in homozygous variants
mie_cands = nb[!(nb$variant_id %in% homozygous$variant_id),]

#find nb variants not in comp_hets
mie_cands = mie_cands[!(mie_cands$variant_id %in% comp_hets$variant_id),]

##########################
#find de novo mutations ##
##########################
#if variant = 1/1 must have one alt from mom and one from dad

#find het mie nb candidates
mie_hom_alt = mie_cands[(mie_cands$gt == "1/1"),]

#check for alts not from either mom or dad
mie_hom_alt[!((mie_hom_alt$alt_id %in% het_dad$alt_id) | (mie_hom_alt$alt_id %in% het_mom$alt_id)),]

dad[(grep("1:162767168:A", dad$alt_id)),]


head(dad)

#if variant 0/0 must have ref from mom and ref from dad


#if variant 0/1 must have one ref from mom and one from dad 








```