---
title: "find_variants_ben"
output: pdf_document
---
# Purpose: 
Pull git from home for the intro. 

#Query
This query was run to find variants in regions of interest: 
```sql
WITH ens AS (
 SELECT DISTINCT chromosome as chr, start, stop, gene_name
   FROM public_hg19.ensembl_genes
  WHERE (gene_name IN ( 'HADH', 'HADHA', 'HADHB', 'ACAA1',
                        'ACAA2', 'EHHADH', 'ECHS1')
     OR gene_name LIKE 'HSD17B%')
    AND chromosome NOT LIKE "H%"
)
SELECT p.*, k.id, k.ref, k.alt, (k.alle_freq * 100) as kav_freqPct, k.alle_cnt as kav_count
  FROM
(SELECT DISTINCT p.sample_id, p.chr, p.pos, p.id,
                 p.ref, p.alt, p.qual, p.filter,
                 ens.gene_name
   FROM ens, p7_ptb.itmi_102_puzzle p
  WHERE p.chr = ens.chr
    AND p.pos >= ens.start AND p.pos <= ens.stop
    AND p.gt IS NOT NULL
) AS p
LEFT JOIN public_hg19.kaviar k
       ON p.chr = k.chromosome
      AND p.pos = k.pos
      AND p.ref = k.ref
      AND p.alt = k.alt
      AND k.alle_freq < .10
```

#Analysis 
Assumption that genes do not span multiple chromosomes. 

```{r}
library(readr)
library(sqldf)

#read in query results
query = read_csv("~/impala_scripts/queries/brady_requests/ben_query.csv")

#split apart data frame by member
mom = query[which(query$member == "M"),]
dad = query[which(query$member == "F"),]
nb = query[which(query$member == "NB"),]

#######################
#find homozygous alt ##
#######################

#subset for zygosity congruent with hom alt mutations
hom_alt_nb = nb[which(nb$gt == "1/1"),]
het_mom = mom[which(mom$gt == "0/1"),]
het_dad = dad[which(dad$gt == "0/1"),]

#merge together to match by gene, chr and position
nb_mom_hom_alt = merge(hom_alt_nb, het_mom, by=c("gene_name", "chr", "pos"))
#get rid of unnecessary columns
nb_mom_hom_alt = nb_mom_hom_alt[,c(1:3,5,9:12,6,7,17)]

#merge with dad
hom_alt = merge(nb_mom_hom_alt, het_dad, by=c("gene_name", "chr", "pos"))

#get rid of unnecessary columns
colnames(hom_alt)
hom_alt = hom_alt[,c(1:11,15)]
colnames(hom_alt) = c("gene_name", "chr", "pos", "rsID", "qual", "filter", "kav_freqPct", 
                      "kav_count", "ref", "nb_alt", "m_alt", "f_alt")

hom_alt_results = hom_alt[which((hom_alt$nb_alt == hom_alt$m_alt) & (hom_alt$nb_alt == hom_alt$f_alt)),]
hom_alt_results

#############################
#find homozygous recessive ##
#############################
hom_rec_nb = nb[which(nb$gt == "0/0"),]

#merge hom rec nb with het mom
hom_rec_nb_mom = merge(hom_rec_nb, het_mom, by=c("gene_name", "chr", "pos"))
#get rid of unnecessary columns
hom_rec_nb_mom = hom_rec_nb_mom[,c(1:3,5,9:12,6,7,17)]

#merge with dad
hom_rec = merge(hom_rec_nb_mom, het_dad, by=c("gene_name", "chr", "pos"))
#get rid of unnecessary columns
hom_rec = hom_rec[,c(1:11,15)]
colnames(hom_rec) = c("gene_name", "chr", "pos", "rsID", "qual", "filter", "kav_freqPct", 
                      "kav_count", "ref", "nb_alt", "m_alt", "f_alt")
hom_rec

#####################
#find compound het ##
#####################

#subset for possible zygosity for compound het
het_nb = nb[which(nb$gt == "0/1"),]
comp_mom = mom[which(mom$gt == "0/1" | mom$gt == "1/1"),]
comp_dad = dad[which(dad$gt == "0/1" | dad$gt == "1/1"),]

#merge het newborn with mom
comp_het_nb_mom = merge(het_nb, comp_mom, by=c("gene_name", "chr"))

#get rid of unnecessary columns
comp_het_nb_mom = comp_het_nb_mom[,c(1,2,9:12,5,4,6:7,15,17,18)]

#merge results with dad
comp_hets = merge(comp_het_nb_mom, comp_dad, by=c("gene_name", "chr"))
#get rid of unnecessary columns
comp_het = comp_hets[,c(1:13,15,17,18)]
colnames(comp_het) = c("gene_name", "chr", "qual", "filter", "kav_freqPct", "kav_count", "nb_rsID", "nb_pos", "nb_ref", 
                       "nb_alt", "m_pos", "m_ref", "m_alt", "f_pos", "f_ref", "f_alt")



head(comp_het)

```