---
title: "ucsc_testing"
output: html_document
---
##Purpose: 
Testing ucsc gene annotation queries. 

##Test plan
The first 100000 rows of  p7_ptb.illumina_variant table were saved as p7dev.illumina_test on impala and run through tests against public_hg19.ucsc to compare results between R and impala queries. 

After initial check to make sure tables in R and impala are both 100,000 rows:

1. Run coding gene annotation query on impala and R 
2. Comapre count of total records returned  
3. Compare number of unique zygosity entries  
4. Compare number of unique variant types returned 
5. Compare count of unique gene names reported   

##Testing Query to find variants in ucsc genes
The following tests were run on p7dev.illumina_test and public_hg19.ucsc tables on impala. 

###Finding variants on impala
The following query was run to count rows returned on the illumina_test and public_hg19.ucsc tables:

```sql 
SELECT COUNT(*)
FROM p7dev.illumina_test as ill, public_hg19.ucsc as ucsc
WHERE ill.filter = "PASS"  
AND ill.chromosome = SUBSTR(ucsc.chrom,4) 
AND ill.position BETWEEN ucsc.txstart AND ucsc.txend  
```
115,769 rows were returned. 

###Downloading variants into R
The following query was run to return results to download for later comparison with R results. Substring was used to account for different format in chromosome name (and this was added to the Needs_Fixed tab of the ill_info spreadshet.)

```sql
SELECT ill.sample_id as ill_sample_id, ill.chromosome as ill_chrom, 
ill.position as ill_pos,ill.ref as ill_ref, ill.alt as ill_alt, ill.id as ill_rsID, 
ucsc.name as ucsc_id, SUBSTR(ucsc.chrom,4) as ucsc_chrom, ucsc.strand as ucsc_strand, 
ucsc.txstart as ucsc_txstart, ucsc.txend as ucsc_txend, ucsc.cdsstart as ucsc_cdsstart, 
ucsc.cdsend as ucsc_cdsend, ucsc.exonstarts as ucsc_exonstart, ucsc.exonends as ucsc_exonend, 
ucsc.proteinid as ucsc_proteinid, ucsc.alignid as ucsc_alignid
FROM p7dev.illumina_test as ill, public_hg19.ucsc as ucsc
WHERE ill.filter = "PASS"  
AND ill.chromosome = SUBSTR(ucsc.chrom,4) 
AND ill.position BETWEEN ucsc.txstart AND ucsc.txend  
```
This file was then read into R to ensure validity of queries used for tesing: 

```{r read_results}
#read in results table
ucsc_results = read.csv("./ucsc/ucsc_query_result.csv", header=TRUE)
head(ucsc_results)
dim(ucsc_results)
```

###Reading data into R for comparison
The subset tables were read into R: 

```{r read_in_subset}
suppressMessages(library(RODBC))

#connect using the DSN name you created on your machine
conn <- odbcConnect("Impala DSN")

ill = sqlFetch(conn, "p7dev.illumina_test")
head(ill)
dim(ill)
ucsc = sqlFetch(conn, "public_hg19.ucsc")
head(ucsc)
dim(ucsc)
```

The following was run in R: 
```{r count_features_R}
suppressMessages(library(GenomicRanges))
suppressMessages(library(IRanges))
suppressMessages(library(TxDb.Hsapiens.UCSC.hg19.knownGene))


##subset for only ill rows that passed filtering
ill_pass = ill[which(ill$filter == "PASS"),]
        
#create GRanges objects for comparison
vars = with(ill_pass, GRanges(seqnames=ill_pass$chromosome, 
                         ranges=IRanges(ill_pass$pos, width=1, 
                                        names=paste(ill_pass$sample_id, ":",ill_pass$ref, ">", ill_pass$alt, sep="")),rsID=ill_pass$id, filter=ill_pass$filter, qual=ill_pass$qual, sample_id=ill_pass$sample_id))

genes = with(ucsc, GRanges(seqnames=as.character(gsub("chr", "", ucsc$chrom)),
                         ranges=IRanges(ucsc$txstart, ucsc$txend,names=ucsc$name),
                         txstart = ucsc$txstart,
                         txend = ucsc$txend,
                         cdsstart = ucsc$cdsstart,
                         cdsend = ucsc$cdsend,
                         exonstart = ucsc$exonstart,
                         exonend = ucsc$exonend,
                         proteinid = ucsc$proteinid
                         ))

##find variants that fall in gene regions
match = suppressWarnings(findOverlaps(vars, genes))

##turn results into a data frame
match.df = data.frame(sample_id = as.character(mcols(vars)$sample_id)[queryHits(match)],
                      chrom = seqnames(vars)[queryHits(match)],
                      pos = as.data.frame(ranges(vars))$start[queryHits(match)],
                      variant = as.character(lapply(strsplit(as.character(names(vars)[queryHits(match)]), ":"), function(x) x[2])),
                      rsId = as.character(mcols(vars)$rsID)[queryHits(match)],
                      gene_name = as.character(names(genes))[subjectHits(match)],
                      txstart = mcols(genes)$txstart[subjectHits(match)],
                      txend = mcols(genes)$txend[subjectHits(match)],
                      cdsstart = mcols(genes)$cdsstart[subjectHits(match)],
                      cdsend = mcols(genes)$cdsend[subjectHits(match)],
                      exonstart = as.character(mcols(genes)$exonstart[subjectHits(match)]),
                      exonend = as.character(mcols(genes)$exonend[subjectHits(match)]),
                      proteinid = mcols(genes)$proteinid[subjectHits(match)],
                      stringsAsFactors=F)
dim(match.df)
#115769 rows, same as impala query                                  
head(match.df)
```

###compare Impala Results file and R generated results
```{r compare_results}
#add variant column to impala results
ucsc_results$variant = paste(ucsc_results$ill_ref, ">", ucsc_results$ill_alt, sep="")
#coerce to same structure as match.df for comparison
ucsc.df = ucsc_results[c(1,2,3,19, 6,7,10:16)]
colnames(ucsc.df) = colnames(match.df)

#order both data frames for matching
ucsc.df = ucsc.df[with(ucsc.df, order(sample_id, chrom, pos, variant, gene_name)),]
match.df = match.df[with(match.df, order(sample_id, chrom, pos, variant, gene_name)),]

#coercing columns to same class for comparison
i = sapply(ucsc.df, is.factor)
ucsc.df[i] = lapply(ucsc.df[i], as.character)

j = sapply(match.df, is.factor)
match.df[j] = lapply(match.df[j], as.character)

#removing rownames for matching
rownames(match.df)= NULL
rownames(ucsc.df) = NULL

#compare data frames
which(match.df != ucsc.df, arr.ind=TRUE)

#0 results returned  = data frames are identical

```

###Count unique protein id's returned
Imapala: 

This following query was run on impala, 776 unique protein id's returned: 
```sql
SELECT COUNT(DISTINCT(ucsc.proteinid))
FROM p7dev.illumina_test as ill, public_hg19.ucsc as ucsc
WHERE ill.filter = "PASS"  
AND ill.chromosome = SUBSTR(ucsc.chrom,4) 
AND ill.position BETWEEN ucsc.txstart AND ucsc.txend  
```

R on impala results: 

The following R query run on the impala results generated above shows that there are 776 unique protein id's returned. 

```{r count_results_features}
str(ucsc_results$ucsc_proteinid)
```

R results: 
776 unique protein id's returned in results, as above. 
```{r count_R_features}
length(unique(ucsc.df$proteinid))

```

###Compare random sample of genes returned
First, we'll grab a random sample of genes from R and examine them on each platform: 
```{r rand_genes}
#sample(unique(match.df$gene_name), 5)
#the above generated the following list, saving as fixed ("uc031qze.1","uc004dhp.3","uc003wvt.4","uc021ujc.2","uc003llx.3")
rand_genes = c("uc031qze.1","uc004dhp.3","uc003wvt.4","uc021ujc.2","uc003llx.3")
```

371rows are returned for these genes in R.  
```{r count_rand_genes}

rand_match = match.df[(match.df$gene_name %in% rand_genes),]
dim(rand_match)
head(rand_match)

```

371 identical genes are also returned from impala results. 

```{r count_rand_impala_results}

match_imp_results = ucsc.df[(ucsc.df$gene_name %in% rand_genes),]
dim(match_imp_results)

##are the gene also the same in the imapla results?
which(rand_match != match_imp_results, arr.ind=TRUE)
#returns 0 mismatches
```
371 identical results are returned on impala: 
```sql
SELECT COUNT(*)
FROM p7dev.illumina_test as ill, public_hg19.ucsc as ucsc
WHERE ill.filter = "PASS"  
AND ill.chromosome = SUBSTR(ucsc.chrom,4) 
AND ill.position BETWEEN ucsc.start AND ucsc.stop 
AND ucsc.name IN ("uc031qze.1","uc004dhp.3","uc003wvt.4","uc021ujc.2","uc003llx.3")
```

###Query to find all variants in ucsc annotated regions
The results of above analysis were used to validate the following query that can be used to determine what regions a variant falls in and for futher annotation in finding amino acid changes:  

```sql 
SELECT ill.sample_id as ill_sample_id, ill.chromosome as ill_chrom, 
ill.position as ill_pos,ill.ref as ill_ref, ill.alt as ill_alt, ill.id as ill_rsID, 
ucsc.name as ucsc_id, SUBSTR(ucsc.chrom,4) as ucsc_chrom, ucsc.strand as ucsc_strand, 
ucsc.txstart as ucsc_txstart, ucsc.txend as ucsc_txend, ucsc.cdsstart as ucsc_cdsstart, 
ucsc.cdsend as ucsc_cdsend, ucsc.exonstarts as ucsc_exonstart, ucsc.exonends as ucsc_exonend, 
ucsc.proteinid as ucsc_proteinid, ucsc.alignid as ucsc_alignid,
CASE 
WHEN ill.position BETWEEN ucsc.cdsstart AND ucsc.cdsend 
THEN 'CDS' 
WHEN ill.position BETWEEN ucsc.txstart AND ucsc.txend
THEN 'transcript'
ELSE NULL
             END AS ucsc_feature
FROM p7dev.illumina_test as ill, public_hg19.ucsc as ucsc
WHERE ill.filter = "PASS"  
AND ill.chromosome = SUBSTR(ucsc.chrom,4) 
AND ill.position BETWEEN ucsc.txstart AND ucsc.txend  
```

##Query to find variants not in ucsc annotated regions
###Impala Results
Results can be used to locate intergenic variants; downstream analysis can be used to find the two flanking genes, and distances between the variants and the flanking genes.

```sql
SELECT ill.* 
FROM
  p7dev.illumina_test as ill
  LEFT JOIN public_hg19.ucsc as ucsc
    ON ill.chromosome = SUBSTR(ucsc.chrom,4) 
    AND ill.position BETWEEN ucsc.txstart AND ucsc.txend
WHERE
  ill.filter = "PASS" 
  AND ucsc.chrom IS NULL
```

This query returned 52,277 rows on impala. 

###Examining random sample of intergenic variants
A random sample of intergenic variants was compared with variants annotated in R using the bioconductor hg19 ensembl database: 

```{r load_intergenic}
intergenic_csv = read.csv("./ucsc/intergenic_query_result.csv", header=TRUE)
head(intergenic_csv)
dim(intergenic_csv)

#pull random sample and assign to variable
rand_ill_intergenic = intergenic_csv[sample(unique(c(intergenic_csv$sample_id)), 5),]
rand_ill_intergenic

suppressMessages(library(VariantAnnotation))
suppressMessages(library(AnnotationHub))
suppressMessages(library(TxDb.Hsapiens.UCSC.hg19.knownGene))

#load referenece genome from ucsc knowngene db
txdb_hg19 <- TxDb.Hsapiens.UCSC.hg19.knownGene


#create granges object from random sample of intergenics
not_genes = with(rand_ill_intergenic, GRanges(seqnames=chromosome, 
                              ranges=IRanges(position, width=1, 
                              names=paste(sample_id, ":",ref, ">", alt, sep="")),
                              rsID = id,
                              sample_id = sample_id))

#chr names don't match
head(seqlevels(txdb_hg19))
seqlevels(not_genes)
#matching chr names
seqlevels(not_genes) <- paste0("chr", seqlevels(not_genes))
not_genes

#assign genome
genome(not_genes) = "hg19"

#confirm that random sample is intergenic
locateVariants(not_genes, txdb_hg19, AllVariants())

```
