---
title: "ucsc_cgi_testing.rmd"
output: md_document
---
###Purpose: 
Testing ucscembl gene annotation queries for cgi variants. 

###Test plan
The first 100,000 rows of  p7_ptb.comgen_variant and public_hg19.ucscembl_genes were subset and saved on impala as p7dev.cgi_test and p7dev.ucscembl_test. 

After an initial check to make sure tables in R and impala are both 100,000 rows:
  
1. Run coding gene annotation query on impala and in R 
2. Comapre count of total records returned  
3. Compare number of unique zygosity entries  
4. Compare number of unique variant types returned 
5. Compare count of unique gene names reported   
6. Compare counts of ucscembl features returned
7. Compare random sample of results between R, impala and bioconductor for hg19

###Finding variants on impala
The following query was run to count the number of anntoated genes returned. Subtracting 1 from cgi coords to match 0-based UCSC until all databses are on the same coordinate system: 
  
```sql 
SELECT COUNT(*)
FROM p7dev.cgi_test as cgi, public_hg19.ucsc as ucsc
WHERE cgi.allele1varquality = "VQHIGH"  
AND cgi.totalreadcount > 30
AND cgi.zygosity = "hom"
AND cgi.chromosome = SUBSTR(ucsc.chrom,4)
AND cgi.start  -1 >= ucsc.txstart 
AND cgi.stop -1 <= ucsc.txend 
```
30,750 rows were returned. 

###Downloading variants into R
The following query was run to return results for later comparison with R results: 
```sql
SELECT cgi.sample_id as cgi_sample_id, cgi.chromosome as cgi_chrom,cgi.start as cgi_start,
cgi.stop as cgi_stop,cgi.reference as cgi_ref,cgi.zygosity as cgi_zygosity,cgi.allele1seq as cgi_alt1,
cgi.allele2seq as cgi_alt2, cgi.allele1varquality as cgi_alt1qual, cgi.vartype as cgi_vartype,
cgi.allele2varquality as cgi_alt2qual, cgi.totalreadcount as cgi_readcount,ucsc.name as ucsc_id, SUBSTR(ucsc.chrom,4) as ucsc_chrom, ucsc.strand as ucsc_strand, ucsc.txstart as ucsc_txstart, 
ucsc.txend as ucsc_txend, ucsc.cdsstart as ucsc_cdsstart, ucsc.cdsend as ucsc_cdsend, 
ucsc.exonstarts as ucsc_exonstart, ucsc.exonends as ucsc_exonend, ucsc.proteinid as ucsc_proteinid, ucsc.alignid as ucsc_alignid,
CASE 
WHEN cgi.start >= ucsc.cdsstart AND cgi.stop <= ucsc.cdsend 
THEN 'CDS' 
WHEN cgi.start >= ucsc.txstart AND cgi.stop <= ucsc.txend
THEN 'transcript'
ELSE NULL
             END AS ucsc_feature
FROM p7dev.cgi_test as cgi, public_hg19.ucsc as ucsc
WHERE cgi.allele1varquality = "VQHIGH"  
AND cgi.totalreadcount > 30
AND cgi.zygosity = "hom"
AND cgi.chromosome = SUBSTR(ucsc.chrom,4)
AND cgi.start -1 >= ucsc.txstart 
AND cgi.stop -1 <= ucsc.txend 
```
This file was then read into R to ucsc to ensure validity of queries used for tesing: 
  
```{r read_results}
#read in results table
ucsc_results = read.csv("./ucsc/ucsc_cgi_query.csv", header=TRUE)
dim(ucsc_results)
head(ucsc_results)
```

###Reading data into R for comparison

The impala tables were read into R: 
  
```{r read_in_subset}
suppressMessages(library(RODBC))

#connect using the DSN name you created on your machine
conn <- odbcConnect("Impala DSN")

cgi = sqlFetch(conn, "p7dev.cgi_test")
head(cgi)
dim(cgi)
ucsc = sqlFetch(conn, "public_hg19.ucsc")
head(ucsc)
dim(ucsc)
```

The following was run in R to find overalps bewteen cgi variants and ucscembl genes: 
```{r count_features_R}
suppressMessages(library(GenomicRanges))
suppressMessages(library(IRanges))

##subset for only cgi rows that have high quality alt1allele and read count > 30
cgi_pass = cgi[which(cgi$allele1varquality == "VQHIGH" & cgi$totalreadcount > 30
                     & cgi$zygosity == "hom"),]

#create GRanges objects for comparison
vars = with(cgi_pass, GRanges(seqnames=chromosome, 
                              ranges=IRanges(start -1, stop -1, 
                              names=paste(sample_id, ":",reference, ">", allele1seq, ":", start, sep="")),zygosity=zygosity, vartype=vartype,
                              read_count = totalreadcount, 
                              sample_id = sample_id))

genes = with(ucsc, GRanges(seqnames=as.character(gsub("chr", "", ucsc$chrom)),
                         ranges=IRanges(ucsc$txstart, ucsc$txend,names=ucsc$name),
                         txstart = ucsc$txstart,
                         txend = ucsc$txend,
                         cdsstart = ucsc$cdsstart,
                         cdsend = ucsc$cdsend,
                         exonstart = ucsc$exonstart,
                         exonend = ucsc$exonend,
                         proteinid = ucsc$proteinid
                         ))

##find variants that fall in gene regions
match = suppressWarnings(findOverlaps(vars, genes))

##turn results into a data frame
match.df = data.frame(sample_id = mcols(vars)$sample_id[queryHits(match)], 
                      chrom = seqnames(vars)[queryHits(match)],
                      start = as.data.frame(ranges(vars))$start[queryHits(match)], 
                      end = as.data.frame(ranges(vars))$end[queryHits(match)],
                      variant = as.character(lapply(strsplit(as.character(names(vars)[queryHits(match)]), ":"), function(x) x[2])), 
                      vartype = as.character(mcols(vars)$vartype)[queryHits(match)],
                      zygosity = as.character(mcols(vars)$zygosity)[queryHits(match)],
                      read_count = as.character(mcols(vars)$read_count)[queryHits(match)],
                      gene_name = as.character(names(genes))[subjectHits(match)],
                      txstart = mcols(genes)$txstart[subjectHits(match)],
                      txend = mcols(genes)$txend[subjectHits(match)],
                      cdsstart = mcols(genes)$cdsstart[subjectHits(match)],
                      cdsend = mcols(genes)$cdsend[subjectHits(match)],
                      exonstart = as.character(mcols(genes)$exonstart[subjectHits(match)]),
                      exonend = as.character(mcols(genes)$exonend[subjectHits(match)]),
                      proteinid = mcols(genes)$proteinid[subjectHits(match)],
                      stringsAsFactors=F)

#80989, 6 more than impala query, let's take a look                              
head(match.df)
dim(match.df)   

```

###Compare Impala Results file and R generated results

```{r compare_results}
#add variant column to impala results
ucsc_results$variant = paste(ucsc_results$cgi_ref, ">", ucsc_results$cgi_alt1, sep="")
#coerce to same structure as match.df for comparison
ucsc.df = ucsc_results[c(1,2,3,4,25,10,6,12,13,16:22)]
colnames(ucsc.df) = colnames(match.df)

#order both data frames for matching
ucsc.df = ucsc.df[with(ucsc.df, order(sample_id, chrom, start, end, variant)),]
match.df = match.df[with(ucsc.df, order(sample_id, chrom, start, end, variant)),]

#coercing columns to same class for comparison
i = sapply(ucsc.df, is.factor)
ucsc.df[i] = lapply(ucsc.df[i], as.character)

j = sapply(match.df, is.factor)
match.df[j] = lapply(match.df[j], as.character)

#removing rownames for matching
rownames(match.df)= NULL
rownames(ucsc.df) = NULL

#compare data frames
which(match.df != ucsc.df, arr.ind=TRUE)

```

No results returned means that the results are identical. 

###Count zygosity
This following query was run on impala to count number of entries in zygosity: 

```sql
SELECT COUNT(cgi.zygosity)
FROM p7dev.cgi_test as cgi, p7dev.ucscembl_test as ucsc
WHERE cgi.allele1varquality = "VQHIGH"  
AND cgi.totalreadcount > 30
AND cgi.chromosome = ucsc.chromosome
AND cgi.start >= ucsc.start 
AND cgi.stop <= ucsc.stop
GROUP BY cgi.zygosity
```

```{r zygosity_table, echo=FALSE}
suppressMessages(library(knitr))
kable(data.frame(zygosity = c("het-ref", "hom"), count=c("64", "18")), 
      align="l", format="pandoc")

```

R on impala results: 
  
The following R query run on the downloaded impala results and generated the same zygosity counts as the imapala query.  

```{r count_results_zygosity}
table(ucsc_results$cgi_zygosity)

```

R results:  

The R generated data was also consistent. 

```{r count_R_zygosity}
table(match.df$zygosity)

```

###Compare variant type counts
This query was run on impala to count number of vartypes returned: 

```sql
SELECT COUNT(cgi.vartype)
FROM p7dev.cgi_test as cgi, p7dev.ucscembl_test as ucsc
WHERE cgi.allele1varquality = "VQHIGH"  
AND cgi.totalreadcount > 30
AND cgi.chromosome = ucsc.chromosome
AND cgi.start >= ucsc.start 
AND cgi.stop <= ucsc.stop
GROUP BY cgi.vartype
```

```{r vartype_table, echo=FALSE}
kable(data.frame(variant_type = c("del", "ins", "snp"), count=c("2", "6", "74")), 
      align="l", format="pandoc")
```

R on impala results:   
  
The impala results were also counted in R.   

```{r count_results_vartype}
table(ucsc_results$cgi_vartype)

```

R generated results: 

Results from the R-generated data are consistent with the impala query:  

```{r count_R_vartyp}
table(match.df$vartype)

```

##Compare count of unique gene names returned
This following query was run on impala to count the number of unique gene names returned: 

```sql
SELECT COUNT(ucsc.gene_name)
FROM p7dev.cgi_test as cgi, p7dev.ucscembl_test as ucsc
WHERE cgi.allele1varquality = "VQHIGH"  
AND cgi.totalreadcount > 30
AND cgi.chromosome = ucsc.chromosome
AND cgi.start >= ucsc.start 
AND cgi.stop <= ucsc.stop
GROUP BY ucsc.gene_name
```

```{r gene_table, echo=FALSE}
kable(data.frame(gene_name = c("C10orf128", "CTNNA3", "FAM170B", "LRRC18", "RP11-298H24.1", "RP11-45P22.2", "RP11-523O18.5", "RP11-523O18.7", "RP11-534L6.3", "VSTM4", "WDFY4"), count=c("4", "13", "1", "4", "3", "1", "1", "1", "1","8", "45")), 
      align="l", format="pandoc")
```

R on impala results:    
  
The impala results were also counted in R. 

```{r count_results_genes}
table(ucsc_results$ucscembl_gene_name)

```

R results:   

Results from the R-generated data are consistent with the impala query.   

```{r count_R_genes}
table(match.df$gene_name)

```

##Compare count of ucscembl features returned
This following query was run on impala to count number of unique gene names returned: 

```sql
SELECT COUNT(ucsc.feature)
FROM p7dev.cgi_test as cgi, p7dev.ucscembl_test as ucsc
WHERE cgi.allele1varquality = "VQHIGH"  
AND cgi.totalreadcount > 30
AND cgi.chromosome = ucsc.chromosome
AND cgi.start >= ucsc.start 
AND cgi.stop <= ucsc.stop
GROUP BY ucsc.feature
```

```{r feature_table, echo=FALSE}
kable(data.frame(gene_name = c("CDS", "exon"), count=c("18", "64")), 
      align="l", format="pandoc")
```

R on impala results:   
  
The impala results were also counted in R.   

```{r count_results_features}
table(ucsc_results$feature)

```

R results:   

Results from the R-generated data are consistent with the impala query. 

```{r count_R_features}
table(match.df$feature)

```

###Compare random sample of genes returned
First, we'll grab a random sample of genes from R and examine them on each platform: 
```{r rand_genes}
#rand_genes = sample(unique(match.df$gene_name), 5)
#the above generated the following list, saving as fixed ("RP11-523O18.7","RP11-298H24.1", "WDFY4","VSTM4", "LRRC18")
rand_genes = c("RP11-523O18.7","RP11-298H24.1", "WDFY4","VSTM4", "LRRC18")
```

How many rows are returned for these genes in R?   
```{r count_rand_genes}

rand_match = match.df[(match.df$gene_name %in% rand_genes),]
dim(rand_match)
head(rand_match)

```

How many of these genes are in impala results? 

```{r count_rand_impala_results}

match_imp_results = ucsc.df[(ucsc.df$gene_name %in% rand_genes),]
dim(match_imp_results)
head(match_imp_results)

##are the gene also the same in the imapla results?
which(rand_match != match_imp_results, arr.ind=TRUE)
#returns 0 mismatches
```

61 results are returned on impala: 

```sql
SELECT COUNT(*)
FROM p7dev.cgi_test as cgi, p7dev.ucscembl_test as ucsc
WHERE cgi.allele1varquality = "VQHIGH"  
AND cgi.totalreadcount > 30
AND cgi.chromosome = ucsc.chromosome
AND cgi.start >= ucsc.start 
AND cgi.stop <= ucsc.stop
AND ucsc.gene_name IN ("RP11-523O18.7","RP11-298H24.1", "WDFY4","VSTM4", "LRRC18")
```

##Query to locate intergenic variants
Results of this query can be used to locate intergenic variants; downstream analysis can be used to find the two flanking genes, and distances between the variants and the flanking genes.

The following query returned 56480 rows and the results were downloaded for analysis in R: 

```sql
SELECT cgi.* 
FROM
p7dev.cgi_test as cgi
LEFT JOIN p7dev.ucscembl_test as ucsc
ON cgi.chromosome = ucsc.chromosome
AND cgi.start >= ucsc.start 
AND cgi.stop <= ucsc.stop
WHERE cgi.allele1varquality = "VQHIGH"  
AND cgi.totalreadcount > 30
AND ucsc.chromosome IS NULL
```
Each of the randomly selected variants was compared with R's annoatation database for hg19 using bioconductor: 

```{r check_intergenic}
#load results from impala
intergenic_csv = read.csv("D:/Documents/GitHub/impala_scripts/queries/testing/ucscembl/intergenic_cgi.csv", header=TRUE)
head(intergenic_csv)

#pull random sample and assign to variable
rand_intergenic = intergenic_csv[sample(unique(c(intergenic_csv$sample_id)), 5),]
rand_intergenic

suppressMessages(library(VariantAnnotation))
suppressMessages(library(AnnotationHub))
suppressMessages(library(TxDb.Hsapiucsc.UCSC.hg19.knownGene))

#load referenece genome from ucsc knowngene db
txdb_hg19 <- TxDb.Hsapiucsc.UCSC.hg19.knownGene


#create granges object from random sample of intergenics
not_genes = with(rand_intergenic, GRanges(seqnames=chromosome, 
                              ranges=IRanges(start, stop, 
                              names=paste(sample_id, ":",reference, ">", allele1seq, sep="")),zygosity=zygosity, vartype=vartype,
                              read_count = totalreadcount, 
                              sample_id = sample_id))
#chr names don't match
head(seqlevels(txdb_hg19))
seqlevels(not_genes)
#matching chr names
seqlevels(not_genes) <- paste0("chr", seqlevels(not_genes))
not_genes

#assign genome
genome(not_genes) = "hg19"

#confirm that random sample is intergenic
locateVariants(not_genes, txdb_hg19, AllVariants())


###testing####
###########################################################
suppressMessages(library(VariantAnnotation))
suppressMessages(library(AnnotationHub))
suppressMessages(library(TxDb.Hsapiucsc.UCSC.hg19.knownGene))

#load referenece genome from ucsc knowngene db
txdb_hg19 <- TxDb.Hsapiucsc.UCSC.hg19.knownGene


##subset for only cgi rows that have high quality alt1allele and read count > 30
cgi_pass = cgi[which(cgi$allele1varquality == "VQHIGH" & cgi$totalreadcount > 30
                     & cgi$zygosity == "hom"),]

#create GRanges objects for comparison
vars = with(cgi_pass, GRanges(seqnames=chromosome, 
                              ranges=IRanges(start -1, stop -1, 
                              names=paste(sample_id, ":",reference, ">", allele1seq, ":", start, sep="")),zygosity=zygosity, vartype=vartype,
                              read_count = totalreadcount, 
                              sample_id = sample_id))

#chr names don't match
head(seqlevels(txdb_hg19))
seqlevels(not_genes)
#matching chr names
seqlevels(not_genes) <- paste0("chr", seqlevels(not_genes))
not_genes

#assign genome
genome(not_genes) = "hg19"

#confirm that random sample is intergenic
locateVariants(not_genes, txdb_hg19, AllVariants())

```
