---
title: "testing_ensmebl.rmd"
output: html_document
---
##Purpose: 
Testing ensembl gene annotation queries. 

##Test plan
A subset of the p7_ptb.illumina_variant table will be created for testing and run through the following tests:  

1. Check that count is the same between impala and table in R  
2. R testing  
  - Check that all chromosomes are equal   
  - Check that vcf position falls between ens start and stop  
  - Examine distribution of ens_feature field  
  - Take random sample and check entries with dbSNP by rsID  
  - Check for variants that matched more than one gene
3. Compare results of above with impala results  
4. In R, subset variants that don't have ensembl annotations  
5. Count variants in non-ensembl set and check for all above  
6. Run non-coding query in impala and check numbers against R  

##Testing query to find all variants in ensembl annotated regions
The results of this query can be used to determine what regions a variant falls in and for futher annotation in finding amino acid changes.

```sql 
SELECT vcf.sample_id as vcf_sample_id, vcf.chromosome as vcf_chrom, vcf.position as vcf_pos,vcf.ref as vcf_ref, vcf.alt as vcf_alt, vcf.id as vcf_rsID, ens.start as ensembl_start, ens.stop as ensembl_end, ens.feature, ens.gene_name as ensembl_gene_name, ens.gene_id as ensembl_geneid, ens.gene_biotype as ensembl_gene_biotype, ens.transcript_name as ensembl_tx_name,ens.transcript_id as ensembl_trans_id, ens.exon_id as ensembl_exonid, ens.strand as ensembl_strand
FROM p7_ptb.illumina_variant as vcf, public_hg19.ensembl_genes as ens  
WHERE vcf.filter = "PASS"  
AND vcf.qual > 100
AND vcf.chromosome = ens.chromosome  
AND vcf.position BETWEEN ens.start AND ens.stop  
```

###Creating vcf subset
The following query was used to subset from p7_ptb.illumina_variant for only variants on chromosome 7 between position 500000 and 600000 that have passed filtering.  

The table was saved in impala as p7dev.sum_test: 

```sql
SELECT vcf.sample_id as sample_id, vcf.position as pos, vcf.id as rsID, vcf.ref as ref, 
vcf.alt as alt, vcf.qual as qual
FROM p7_ptb.illumina_variant as vcf
WHERE vcf.chromosome LIKE "7"
AND vcf.filter = "PASS"
```
Reading the vcf subset into R: 

```{r read_in_subset}
library(RODBC)

#connect using the DSN name you created on your machine
conn <- odbcConnect("Impala DSN")

vcf = sqlFetch(conn, "p7dev.sum_test")
head(vcf)
```

###Counting vcf subset 
The following query resulted in 572490 records on impala:  
```sql
SELECT count(*)
FROM p7_ptb.illumina_variant as vcf
WHERE vcf.filter = "PASS"
AND vcf.chromosome LIKE "7"
AND vcf.position BETWEEN 500000 AND 600000
```

And in R: 
```{r count_subset}
dim(vcf)
```
Counts match in both tables. 

###Read in ensembl table
```{r read_in_ensembl}
ensembl = sqlFetch(conn, "public_hg19.ensembl_genes")
head(ensembl)
dim(ensembl)
```

###Check that chromosomes are equal between two tables
```{r check_chromosomes}
dim(ensembl[which(ensembl$vcf_chrom != ensembl$ens_chrom)])
```

###Check that vcf position falls between ens start and stop
```{r check_pos}




```

###Examine distribution of ens_feature field

```{r check_features}




```

###Take random sample and check entries with dbSNP by rsID 

```{r check_dbSNP}




```

###Check for variants that matched more than one gene
```{r check_variant_to_gene}




```

##Query to find variants not in ensembl annotated regions
Results can be used to locate intergenic variants; downstream analysis can be used to find the two flanking genes, and distances between the variants and the flanking genes.

```sql
SELECT vcf.sample_id as vcf_sample_id, vcf.chromosome as vcf_chrom, vcf.position as vcf_pos, vcf.ref as vcf_ref, vcf.alt as vcf_alt, vcf.id as vcf_rsID, ens.start as ensembl_start, ens.stop as ensembl_end, ens.feature, ens.gene_name as ensembl_gene_name, ens.gene_id as ensembl_geneid, ens.gene_biotype as ensembl_gene_biotype, ens.transcript_name as ensembl_tx_name, ens.transcript_id as ensembl_trans_id, ens.exon_id as ensembl_exonid, ens.strand as ensembl_strand
FROM p7_ptb.illumina_variant as vcf, public_hg19.ensembl_genes as ens
WHERE vcf.filter = "PASS"
AND vcf.qual > 100
AND vcf.chromosome = ens.chromosome
AND vcf.position NOT BETWEEN ens.start AND ens.stop
```
