---
title: "ClinVar results and annotation"
output: pdf_document
---
This analysis was performed on the following trios to locate potentially pathogenic variants using Clinvar (version 2/5/15): 

-	102-07005  
-	102-00997 
-	102-00996  
-	102-02001  
-	102-00961  
-	102-07009  
-	102-00932  
-	102-00979  
-	102-00974  
-	102-01532  

Each variant found was annotated with ClinVar pathogenicity ratings and Kaviar allele frequency using the following parameters: 
- Non-conflicted ClinVar clinical significance rating of 4 or 5, but never 2 or 3
- Homozygous for alternate allele

```sql 
WITH clin as (
  SELECT bv.sample_id, bv.chr, bv.pos, bv.id as rsID, bv.ref, bv.alt, bv.qual, bv.filter,
  bv.gt, clin.gene, clin.accession, clin.clin_acc, clin.clin_sig, clin.clin_hgvs, clin.clin_dbn
  FROM p7_platform.brady_variant as bv, public_hg19.clinvar AS clin
  WHERE (bv.sample_id LIKE '%01' OR bv.sample_id LIKE '%02' OR bv.sample_id LIKE '%03')
  AND (clin_sig NOT REGEXP '3|2[^5]|2$'
  AND clin_sig REGEXP '4|[^25]5|^5')
  AND clin.chrom = bv.chr
  AND clin.ref = bv.ref
  AND clin.alt = bv.alt
  AND clin.pos = bv.pos
  AND (bv.gt <> '0/0' OR bv.gt <> '0')
  )
  
SELECT clin.*, (kav.alle_freq * 100) as kav_freq_pct
   FROM clin
   LEFT JOIN public_hg19.kaviar kav
   ON clin.rsid = kav.id
   AND clin.chr = kav.chromosome
   AND clin.pos = kav.pos
   AND clin.ref = kav.ref
   AND clin.alt = kav.alt
```

The results were then joined with the itmi_mapping table to provide information on subject's gender, age and country of origin.

```sql
SELECT bv.*, mp.member, mp.age, mp.country
FROM users_selasady.clinvar_kav_homalt bv, p7_platform.itmi_mapping mp
WHERE bv.sample_id = mp.subject_id
```

```{r}
library(readr)
library(readxl)
library(plyr)

#read in results of clin annotation
clin = read_csv("./clin_results.csv")

#########################
# Add allele frequency ##
#########################
#add column of unique variant id for matching
clin$var_id = paste(clin$chr, clin$pos, clin$ref, clin$alt, sep=":")

#count the number of times each variant occurs
allele_count = ddply(clin, .(var_id), summarise, 
                     count = length(var_id)) 

#add counts column to clinvar results
clin$var_count = as.integer(allele_count[match(clin$var_id, allele_count$var_id),]$count)

#if not a sex chromosome and hom-alt, allele frequency equals ((var_count * 2)/(nsamples * 2) * 100)                     
clin[(grep("X|Y", clin$chr, invert=TRUE)),]$data_freq = ((clin[(grep("X|Y", clin$chr, invert=TRUE)),]$var_count * 2)/(length(unique(clin[(grep("X|Y", clin$chr, invert=TRUE)),]$sample_id))))*100

#if not a sex chromosome and het, allele frequency equals ((var_count)/(nsamples * 2) * 100)                     
clin[(grep("X|Y", clin$chr, invert=TRUE)),]$data_freq = ((clin[(grep("X|Y", clin$chr, invert=TRUE)),]$var_count)/(length(unique(clin[(grep("X|Y", clin$chr, invert=TRUE)),]$sample_id))))*100

#if a sex chromosome, allele frequency equals ((var_count)/(nsamples) * 100)                     
clin[(grep("X|Y", clin$chr)),]$data_freq = ((clin[(grep("X|Y", clin$chr)),]$var_count)/(length(unique(clin[(grep("X|Y", clin$chr)),]$sample_id))))*100
  
##############################
# Add functional annotation ##
##############################
library(VariantAnnotation)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(org.Hs.eg.db)  
library(BSgenome.Hsapiens.UCSC.hg19)

#Pull out poisitional and variant info
clin_range <- data.frame(clin$rsid, clin$chr, clin$pos, clin$ref, clin$alt)
colnames(clin_range) <- c("rsid", "chr", "pos", "ref", "alt")
clin_range$pos = as.numeric(as.character(clin_range$pos))
clin_range$chr = paste("chr", clin_range$chr, sep="")

#create granges object
clin_targets <- with(clin_range,
                VRanges( seqnames = Rle(chr),
                         ranges   = IRanges(pos, end=pos, names=rsid),
                         ref      = clin$ref,
                         alt      = clin$alt,
                         sampleNames = clin$sample_id
                         ) )

#annotate with prediction of coding changes
setMethod("nchar", "ANY", base::nchar) #fixes error in latest version of rstudio
altallele = DNAStringSet(alt(clin_targets)) #convert alts into DNAStringSet for vranges object
alt_length = elementLengths(altallele) #add length of variants
coding <- predictCoding(clin_targets, TxDb.Hsapiens.UCSC.hg19.knownGene, Hsapiens, altallele) #predict coding

#convert coding results to dataframe for output to file
names(coding) = NULL
coding_out = as.data.frame(coding)
coding_out$names = names(clin_targets)[coding_out$QUERYID]
#remove "chr" from chromosome name for matching
coding_out$chr = lapply(strsplit(as.character(coding_out$seqnames), "chr"), function(x) x[2])

#add code for matching up tables
coding_out$var_code = paste(coding_out$sampleNames, coding_out$chr, coding_out$start, coding_out$ref, coding_out$alt, sep=":")
#subset coding_out to only desired fields
coding_out = coding_out[,c(28, 18,19,21:25)]
#keep only unique variant rows
coding_out = coding_out[!duplicated(coding_out$var_code),]

#annotate with location on reference (ucsc known genes)
loc <- locateVariants(clin_targets, TxDb.Hsapiens.UCSC.hg19.knownGene, AllVariants())
names(loc) <- NULL
loc_out <- as.data.frame(loc)
loc_out$names <- names(clin_targets)[ loc_out$QUERYID ]
loc_out$chr = lapply(strsplit(as.character(loc_out$seqnames), "chr"), function(x) x[2])
loc_out$pos_id = paste(loc_out$chr, loc_out$start, sep=":")

#subset loc_out to only desired fields
loc_out = loc_out[,c(17,6,11)]
#keep only rows with unique pos_id
loc_out = loc_out[!duplicated(loc_out$pos_id),]

#combine results 
clin$var_code = paste(clin$sample_id, clin$chr, clin$pos, clin$ref, clin$alt, sep=":")
clin$pos_id = paste(clin$chr, clin$pos, sep=":")

clin_coding = merge(clin, coding_out, by="var_code", all.x=TRUE)
clin_all = merge(clin_coding, loc_out, by="pos_id", all.x=TRUE)

#remove extraneous columns
clin_annot = clin_all[,c(3:18,20:29)]
colnames(clin_annot) = c("sample_id", "chr", "pos", "rsid", "ref", "alt", "qual", "filter", "gt", "gene", "accession", "clin_acc", "clin_sig", "clin_hgvs", "clin_dbn", "kav_freq", "var_count", "data_freq", "txid", "cdsid", "consequence", "refcodon", "varcodon", "refaa", "varaa", "location")

#split into separate files, as requested
library(dplyr)
clin_hom_parents = filter(clin_annot,  grepl("01$|02$", clin_annot$sample_id), (gt == '1'| gt == '1/1'))
clin_het_parents = filter(clin_annot,  grepl("01$|02$", clin_annot$sample_id), gt == '0/1')
clin_hom_nb = filter(clin_annot,  grepl("03$", clin_annot$sample_id), (gt == '1'| gt == '1/1'))
clin_het_nb = filter(clin_annot,  grepl("03$", clin_annot$sample_id), gt == '0/1')

#does this all add up? 
dim(clin_het_nb)[1] + dim(clin_hom_nb)[1] + dim(clin_hom_parents)[1] + dim(clin_het_parents)
#yes

#save each file as tsv
write.table(clin_hom_parents, file="./clin_hom_parents.tsv", row.names = FALSE, quote = FALSE, sep = "\t")
write.table(clin_het_parents, file="./clin_het_parents.tsv", row.names = FALSE, quote = FALSE, sep = "\t")
write.table(clin_hom_nb, file="./clin_hom_nb.tsv", row.names = FALSE, quote = FALSE, sep = "\t")
write.table(clin_het_nb, file="./clin_het_nb.tsv", row.names = FALSE, quote = FALSE, sep = "\t")

```

Next I ran the following query to find genes in the acmg list: 

```sql
WITH clin AS
(SELECT clin.chrom, clin.pos, clin.ref, clin.alt, clin.id AS rsID, clin.gene, clin.accession, 
 clin.clin_sig, clin.clin_hgvs, clin.clin_dbn, acmg.phenotype AS acmg_phen, acmg.age_onset AS acmg_age_onset,
 acmg.inheritance as acmg_moi, acmg.variants_to_report as acmg_pathogenicity
FROM public_hg19.clinvar AS clin
JOIN users_selasady.acmg_genes AS acmg
ON clin.gene = acmg.gene
WHERE (clin_sig NOT REGEXP '3|2[^5]|2$'
  AND clin_sig REGEXP '4|[^25]5|^5')
)

SELECT bv.sample_id, bv.chr, bv.pos, bv.id as rsID, bv.ref, bv.alt, bv.qual, bv.filter,
bv.gt, clin.gene, clin.accession, clin.clin_sig, clin.clin_hgvs, clin.clin_dbn,
clin.acmg_phen, clin.acmg_age_onset, clin.acmg_moi
FROM p7_platform.brady_variant as bv, clin
WHERE (bv.sample_id LIKE '%03'
    OR bv.sample_id LIKE '%01'
    OR bv.sample_id LIKE '%02')
AND clin.chrom = bv.chr
AND clin.ref = bv.ref
AND clin.alt = bv.alt
AND clin.pos = bv.pos
```
Kaviar annotation and gender/age/country were added as follows: 

```sql
SELECT acmg.*, (kav.alle_freq * 100) as kav_freq
FROM acmg_results acmg, public_hg19.kaviar kav
WHERE acmg.rsid = kav.id
   AND acmg.chr = kav.chromosome
   AND acmg.pos = kav.pos
   AND acmg.ref = kav.ref
   AND acmg.alt = kav.alt
```
I then read the results into R to add protein change predictions. Calculating data set MAF on five variants did not seem necessary.

```{r}
#read in file
acmg = read_csv("./acmg_results.csv")

#convert to vranges object and add annotation
#Pull out poisitional and variant info
acmg_range <- data.frame(acmg$rsid, acmg$chr, acmg$pos, acmg$ref, acmg$alt)
colnames(acmg_range) <- c("rsid", "chr", "pos", "ref", "alt")
acmg_range$pos = as.numeric(as.character(acmg_range$pos))
acmg_range$chr = paste("chr", acmg_range$chr, sep="")

#create granges object
acmg_targets <- with(acmg_range,
                VRanges( seqnames = Rle(chr),
                         ranges   = IRanges(pos, end=pos, names=rsid),
                         ref      = acmg$ref,
                         alt      = acmg$alt,
                         sampleNames = acmg$sample_id
                         ) )

#annotate with prediction of coding changes
acmg_altallele = DNAStringSet(alt(acmg_targets))
acmg_alt_length = elementLengths(acmg_altallele)
acmg_coding <- predictCoding(acmg_targets, TxDb.Hsapiens.UCSC.hg19.knownGene, Hsapiens, acmg_altallele)

#convert coding results to df
names(acmg_coding) = NULL
acmg_coding_out = as.data.frame(acmg_coding)
acmg_coding_out$names = names(acmg_targets)[acmg_coding_out$QUERYID]
acmg_coding_out$chr = lapply(strsplit(as.character(acmg_coding_out$seqnames), "chr"), function(x) x[2])

#add code for matching up tables
acmg_coding_out$var_code = paste(acmg_coding_out$sampleNames, acmg_coding_out$chr, acmg_coding_out$start, acmg_coding_out$ref, acmg_coding_out$alt, sep=":")
#subset coding_out to only desired fields
acmg_coding_out = acmg_coding_out[,c(27, 18,19,21:25)]

#annotate with location on reference (ucsc)
acmg_loc <- locateVariants(acmg_targets, TxDb.Hsapiens.UCSC.hg19.knownGene, AllVariants())
names(acmg_loc) <- NULL
acmg_loc_out <- as.data.frame(acmg_loc)
acmg_loc_out$names <- names(acmg_targets)[ acmg_loc_out$QUERYID ]
acmg_loc_out$chr = lapply(strsplit(as.character(acmg_loc_out$seqnames), "chr"), function(x) x[2])
acmg_loc_out$pos_id = paste(acmg_loc_out$chr, acmg_loc_out$start, sep=":")

#subset loc_out to only desired fields
acmg_loc_out = acmg_loc_out[,c(17,6,11)]

#combine results 
acmg$var_code = paste(acmg$sample_id, acmg$chr, acmg$pos, acmg$ref, acmg$alt, sep=":")
acmg$pos_id = paste(acmg$chr, acmg$pos, sep=":")

acmg_coding = merge(acmg, coding_out, by="var_code", all.x=TRUE)
acmg_all = merge(acmg_coding, loc_out, by="pos_id", all.x=TRUE)

#remove extraneous columns
acmg_annot = acmg_all[,c(3:27)]
colnames(acmg_annot) = c("sample_id", "chr", "pos", "rsid", "ref", "alt", "qual", "filter", "gt", "gene", "accession", "clin_acc", "clin_sig", "clin_hgvs", "clin_dbn", "kav_freq", "txid", "cdsid", "consequence", "refcodon", "varcodon", "refaa", "varaa", "location")

#save results
write.csv(acmg_annot, file="./acmg_analysis_071415.csv", row.names = FALSE)
```
